### PR TITLE
Comparing python objects (such as list/tuple) with `mlx.core.array`

### PR DESCRIPTION
**Pull request by amirhossein-razlighi**
## Proposed changes

This closes issue #866. When trying to compare python list or tuple with an mlx.core.array, mlx throws exception. Right now, regarding the comments in the related issue, mlx returns False (following the way it is done in PyTorch). An example is as follows:
```
import mlx.core as mx

a = mx.array([1, 2, 3])
a == [1, 2, 3] # False
a == mx.array([1,2,3]) # True
a != [1, 2, 3] # True
a == (1, 2) # False
```

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

**Comment by amirhossein-razlighi** [comment]
I'm sorry. I accidentally closed [previous PR](https://github.com/ml-explore/mlx/pull/886#issue-2204299594) regarding to a hard reset of my main branch, and then I couldn't update that PR anymore.

**Comment by amirhossein-razlighi** [comment]
In continue of your [latest review](https://github.com/ml-explore/mlx/pull/886#discussion_r1541844727):

I swapped the hasattr and isinstance. Also, why should we add else { return true; }? Because if it's not each of the options mentioned above (bool/int/float/complex/array) it is not convertible to array and thus, we should simply return False.
Also, I think we need a better name for this function. Because, instances like list and tuple are actually convertible to mlx.core.array. They are not comparable to arrays. Don't you agree? This name might be a little misleading in further developments

File: python/src/utils.h [hunk 1]
```diff
@@ -44,6 +44,23 @@ inline array to_array_with_accessor(nb::object obj) {
...
+    return true;
+  } else if (auto pv = std::get_if<nb::int_>(&v); pv) {
+    return true;
+  } else if (auto pv = std::get_if<nb::float_>(&v); pv) {
+    return true;
+  } else if (auto pv = std::get_if<std::complex<float>>(&v); pv) {
+    return true;
+  } else if (auto pv = std::get_if<nb::object>(&v); pv) {
+    return nb::isinstance<array>(*pv) || nb::hasattr(*pv, "__mlx_array__");
+  }
```
Comment (awni) [inline python/src/utils.h:60]: What I mean is that you replace this entire function with:
```c++
if (auto pv = std::get_if<nb::object>(&v); pv) {
    return nb::isinstance<array>(*pv) || nb::hasattr(*pv, "__mlx_array__");
} else {
    return true;
}
```

The only time it's not an object is when it's a python bool, int or complex.

Comment (amirhossein-razlighi) [reply python/src/utils.h]: Done ✅

File: ACKNOWLEDGMENTS.md [hunk 2]
```diff
@@ -15,7 +15,7 @@ MLX was developed with contributions from the following individuals:
 - Hinrik Snær Guðmundsson: Added `atleast_1d`, `atleast_2d`, `atleast_3d` ops.
 - Luca Arnaboldi: Added `Ceil` and `Floor` ops; implemented pickling, copy and deepcopy for mlx arrays.
 - Brian Keene & Atila Orhon, with Argmax Inc.: Added `fast.scaled_dot_product_attention`
-- AmirHossein Razlighi: Added chaining support for some of the ops in `nn.Module`.
+- AmirHossein Razlighi: Added chaining support for some of the ops in `nn.Module`, Some Debug and Enhancements in `mlx.core.array`
```
Comment (awni) [inline ACKNOWLEDGMENTS.md:18]:
```suggestion
- AmirHossein Razlighi: Added chaining support for some of the ops in `nn.Module`. Comparisons works for non array objects in `mlx.core.array`.
```

**Review by awni — APPROVED** [review]
LGTM, thanks!

**Pull request merged**